File: whatsapp_sender.js
// whatsapp_sender.js
// Sends all unsent notices to WhatsApp channel — caption + file attachment

const { Client, LocalAuth, MessageMedia } = require("whatsapp-web.js");
const qrcode = require("qrcode-terminal");
const sqlite3 = require("better-sqlite3");
const path = require("path");
const fs = require("fs");

const CHANNEL_ID = process.env.WHATSAPP_CHANNEL_ID;
const DB_PATH = path.join(__dirname, "database.db");
const db = new sqlite3(DB_PATH);

function getUnsentNotices() {
  return db.prepare("SELECT * FROM notices WHERE sent_to_whatsapp = 0").all();
}

function markAsSent(id) {
  db.prepare("UPDATE notices SET sent_to_whatsapp = 1 WHERE id = ?").run(id);
}

function buildCaption(notice) {
  const { title, url, file_type, date_on_site } = notice;
  const lines = [
    `📢 *New Notice*`,
    ``,
    `📌 *${title}*`,
    `🕐 ${date_on_site}`,
    `🔗 ${url}`,
  ];
  return lines.join("\n");
}

async function sendNotices(client) {
  const unsent = getUnsentNotices();

  if (unsent.length === 0) {
    console.log("📭 No new notices to send.");
    return;
  }

  console.log(`📤 Sending ${unsent.length} notice(s) to WhatsApp...`);

  for (const notice of unsent) {
    const { id, file_path, file_type } = notice;
    const caption = buildCaption(notice);

    try {
      const ext = path.extname(file_path || "").toLowerCase();
      const fileExists = file_path && fs.existsSync(file_path);

      if (fileExists && ext === ".txt") {
        // Send text content inline
        const content = fs.readFileSync(file_path, "utf-8").slice(0, 3000);
        await client.sendMessage(CHANNEL_ID, `${caption}\n\n─────────────\n${content}`);

      } else if (fileExists) {
        // Send file (PDF, JPG, DOCX) with caption
        const media = MessageMedia.fromFilePath(file_path);
        await client.sendMessage(CHANNEL_ID, media, { caption });

      } else {
        // No file — send caption + link only
        await client.sendMessage(CHANNEL_ID, caption);
      }

      markAsSent(id);
      console.log(`✅ Sent: ${notice.title}`);

      // Small delay to avoid spam detection
      await new Promise((r) => setTimeout(r, 2500));

    } catch (err) {
      console.error(`❌ Failed "${notice.title}": ${err.message}`);
    }
  }

  console.log("✅ All done.");
}

// ── WhatsApp Client ───────────────────────────────────────────────────────────
const client = new Client({
  authStrategy: new LocalAuth({ dataPath: ".wwebjs_auth" }),
  puppeteer: {
    args: [
      "--no-sandbox",
      "--disable-setuid-sandbox",
      "--disable-dev-shm-usage",
      "--disable-gpu",
    ],
    headless: true,
  },
});

client.on("qr", (qr) => {
  console.log("\n📱 Scan this QR code:\n");
  qrcode.generate(qr, { small: true });
});

client.on("ready", async () => {
  console.log("✅ WhatsApp ready!");
  await sendNotices(client);
  process.exit(0);
});

client.on("auth_failure", (msg) => {
  console.error("❌ Auth failed:", msg);
  process.exit(1);
});

client.initialize();

File: whatsapp_sender.py
"""
whatsapp_sender.py
Sends unsent notices to a WhatsApp group using Green API (free tier).
No Node.js, no puppeteer, no session files needed.
"""

import os
import requests
from pathlib import Path
from database import get_unsent_notices, mark_as_sent

INSTANCE_ID = os.getenv("GREEN_API_INSTANCE")
API_TOKEN   = os.getenv("GREEN_API_TOKEN")
GROUP_ID    = os.getenv("WHATSAPP_GROUP_ID")   # e.g. 120363XXXXXXXXXX@g.us
BASE_URL    = f"https://api.green-api.com/waInstance{INSTANCE_ID}"

SUPPORTED_EXTENSIONS = {".pdf", ".jpg", ".jpeg", ".png", ".docx", ".doc"}


def send_message(text):
    """Send a plain text message."""
    url = f"{BASE_URL}/sendMessage/{API_TOKEN}"
    payload = {
        "chatId": GROUP_ID,
        "message": text
    }
    resp = requests.post(url, json=payload, timeout=30)
    resp.raise_for_status()
    return resp.json()


def send_file(file_path, caption=""):
    """Send a file (PDF, image, docx) with caption."""
    url = f"{BASE_URL}/sendFileByUpload/{API_TOKEN}"
    with open(file_path, "rb") as f:
        filename = Path(file_path).name
        resp = requests.post(
            url,
            data={"chatId": GROUP_ID, "caption": caption},
            files={"file": (filename, f)},
            timeout=60
        )
    resp.raise_for_status()
    return resp.json()


def build_caption(notice):
    """Build a clean WhatsApp caption from a notice row."""
    id_, title, url, file_path, file_type, date_on_site, downloaded_at, _ = notice

    # Remove trailing "Download" from titles scraped from the site
    clean_title = title.replace("Download", "").strip()

    lines = [
        "📢 *New Notice*",
        "",
        f"📌 *{clean_title}*",
        f"🕐 {date_on_site}",
        f"🔗 {url}",
    ]
    return "\n".join(lines)


def send_notices():
    unsent = get_unsent_notices()

    if not unsent:
        print("📭 No new notices to send.")
        return

    print(f"📤 Sending {len(unsent)} notice(s) to WhatsApp...")

    for notice in unsent:
        id_ = notice[0]
        file_path = notice[3]
        caption = build_caption(notice)

        try:
            ext = Path(file_path).suffix.lower() if file_path else ""
            file_exists = file_path and os.path.exists(file_path)

            if file_exists and ext == ".txt":
                # Send text content inline
                content = open(file_path, encoding="utf-8").read()[:3000]
                send_message(f"{caption}\n\n─────────────\n{content}")

            elif file_exists and ext in SUPPORTED_EXTENSIONS:
                # Send file with caption
                send_file(file_path, caption=caption)

            else:
                # No file or unsupported type — send caption + link only
                send_message(caption)

            mark_as_sent(id_)
            print(f"  ✅ Sent: {notice[1][:60]}")

            # Small delay between messages
            import time
            time.sleep(2)

        except Exception as e:
            print(f"  ❌ Failed '{notice[1][:40]}': {e}")

    print("✅ All notices sent.")


if __name__ == "__main__":
    send_notices()
